src/Mod/Path/PathScripts/PathWaterline.py

```diff
@@ -1725,6 +1756,525 @@ def _loopToGcode(self, obj, layDep, loop):
+                mask.subtract( void_mask )
+                op.exactShift( mask, -max(obj.BoundaryAdjustment.Value, op.radius))
+            else:
+                mask.shiftBorder( -obj.BoundaryAdjustment.Value )
+                if not void_mask is None:
+                    op.exactShift( void_mask, op.radius )
+                    mask.subtract( void_mask )
+        else:
+            mask.subtract( void_mask )
+            mask.shiftBorder( -obj.BoundaryAdjustment.Value )
+        # Add one row of cells more for gray zone
+        mask.highlightBorders(new_state=2)
+        mask.highlightBorders(threshold=2, new_state=2)
+
+        for m in range(0, len(JOB.Model.Group)):
+            if m != mdlIdx:
+                op.raiseModel( JOB.Model.Group[m], common = base_only )
+        op.cleanupCommonMap()
+        PathLog.info("LevelMap creation %.3f s" % (time.time() - t0))
+
+        t0 = time.time()
+        op.applyTool( self.tool )
+        PathLog.info("The tool application %.3f s" % (time.time() - t0))
+
+        # Extract layers per depthparams
+        layTime = time.time()
+        prevLayDep = obj.StartDepth.Value
+        contourMap = None
+        previousMap = None
+        halfMap    = None
+        backMap    = None
+        all_contours = []     # list of traces, each trace is a list of nodes
+        all_internals = []
+        max_dist_m = (max(self.horizFeed * self.horizRapid, 0.1) /
+                      max(self.horizRapid - self.horizFeed, 0.01) * 2 /
+                      max(self.vertRapid, 0.01)
+                     )
+        max_dist_c = (obj.StepDown.Value * self.vertRapid / max(self.vertFeed, 0.01) +
+                       obj.SafeHeight.Value)
+        step_over = max(4 * op.levelMap.sampleInterval,
+                        self.radius * (float(obj.StepOver) / 50.0) -
+                        op.levelMap.sampleInterval)
+
+        for layDep in depthparams:
+            max_distance = (max_dist_c - layDep) * max_dist_m
+
+            contourMap = op.getContourMap(layDep, obj.DepthOffset.Value,
+                                          out = contourMap)
+
+            traces = []   # list of traces, each trace is a list of nodes
+            contours = None
+            half_map_traces = []
+            while True:  # Collect traces of the single layer
+                if obj.LayerMode == "Single-pass":
+                    contourMap.highlightBorders()
+                    op.excludeCommonFrom(contourMap)
+                    contourMap.or_not( mask )  # simulate material around the work area
+                else:
+                    contourMap.or_not( mask )  # simulate material around the work area
+                    op.excludeCommonFrom(contourMap)
+                    contourMap.highlightBorders()
+
+                # Extract waterline
+                # contourMap.dump("drill_05_%i_%.1f_%i.txt" % (mdlIdx, layDep, len(traces)))
+                nothing = True
+                while True:
```

Comment:
Seems like the only way out of the loop is the break statement.
Couldn't you do
while len(nodes) > 0:  
   ...